Auth using before_request decorator

RenaudLN opened the pull request:
This PR changes the behaviour of Auth (and BasicAuth) to work on every single route by default. Until now, only the routes that were created before the instantiation of Auth were protected.

To ensure every route goes through the auth check, a `before_request` decorator is used.

Bonus: Users can whitelist endpoints by marking them as public via Flask's config using the `PUBLIC_ROUTES` key. The public routes should follow the same format as regular Flask routes.

alexcjohnson commented:
I like this a lot! I wonder if we can (optionally) combine the public routes into the constructor, ie instead of two calls:
```py
add_public_routes(app, ["/home", "/user/<user_id>/public"])
BasicAuth(app, USERS)
```
you could write it as:
```py
BasicAuth(app, USERS, public_routes=["/home", "/user/<user_id>/public"])
```

alexcjohnson commented on dash_auth/auth.py:
```diff
@@ -1,29 +1,45 @@
 
 class Auth(ABC):
-    def __init__(self, app, authorization_hook=None, _overwrite_index=True):
+    def __init__(self, app: Dash, **_kwargs):
```
We probably don't want to silently discard `**_kwargs`, can we just take this out and have it throw a `TypeError: Auth got an unexpected keyword argument <whatever>`? Seems like `authorization_hook` hasn't been used since we removed `PlotlyAuth`, and I don't see a use of `_overwrite_index` even there.

RenaudLN commented:
Good idea, added `public_routes` to the constructor (which under the hood just calls `add_public_routes`).
Also updated the test to check that both paths work.

fkuschel commented:
Hey Renaud,

I was looking exactly for the new feature "integration of public routes" for my multi page dash app. 
I installed your latest branch and somehow the public routes pages still require a password.

Would you mind having a short look into this minimal example? Do you see something wrong here?

`
import dash_auth
from dash import Dash, html, dcc
import dash_bootstrap_components as dbc


app = Dash(__name__)
VALID_USERNAME_PASSWORD_PAIRS = {'fkl': 'fkl'}

dash_auth.BasicAuth(app, VALID_USERNAME_PASSWORD_PAIRS,
                    public_routes=["/"])

app.layout = dbc.Container([
    dcc.Store(id="table_edits"),
    dbc.Row([
        dbc.Col(html.H2(["Test"]), width=12, xl={"size": 4})])])

if __name__ == "__main__":
    app.run_server()
`
Thanks in advance :) 
Best regards 
Florian

RenaudLN commented:
Yes you are right it actually doesn't work in its current state... The issues are:

* All the dash dependencies / layout route etc. need to be whitelisted, that can be done easily
* Some or all the assets need to be whitelisted, for js scripts and css it's fine to whitelist them all but for other assets it might be an issue to make them public (e.g. hosting some static documentation)
* Dash's callback all use a single `/_dash-update-component`, that is also the case for the navigation of multi-page apps. There might be workarounds to catch the callback that does the navigation but the issue would then be to whitelist the callbacks that are accessible on those public routes. Maybe we could use the `request.referrer` to decide whether a callback is allowed but I have a feeling this is not safe as someone on a public route could do a post request to `/_dash-update-component` with the right set of arguments and get the data they want.

Given the above, I think I will remove the public routes part of the PR to keep only the change to a `@before_request`, this will give the benefit of protecting all routes, even ones added after the auth is instantiated.

@alexcjohnson happy to have your thoughts on this or if you see some workarounds I may not have thought about.

RenaudLN commented:
@fkuschel I gave it a bit more thought and I believe I have a working solution. It does require to mark all callbacks on public routes as `public_callbacks` but I don't see any other way around it.

Checkout the updated Readme for usage and example in a multi-page app.

fkuschel commented:
@RenaudLN, 
for me this approach works perfectly! 

Thanks a lot!

olivier-lacroix commented on dash_auth/public_routes.py:
```diff
@@ -0,0 +1,99 @@
+from werkzeug.routing import Map, MapAdapter, Rule
+
+# Add PUBLIC_ROUTES in the default Flask config
+DASH_PUBLIC_ASSETS_EXTENSIONS = "js,css,svg,png,webp,gif"
```
@RenaudLN , I am not sure we should make assets public. It would be reasonable to me that routes for selected assets (or all assets) have to be whitelisted manually

RenaudLN replied:
Asking people to whitelist their custom js scripts and stylesheets seems overkill to me. Not 100% sure about images so maybe I could drop those.

olivier-lacroix commented on dash_auth/public_routes.py:
```diff
@@ -0,0 +1,99 @@
+        "DASH_PUBLIC_ASSETS_EXTENSIONS",
+        DASH_PUBLIC_ASSETS_EXTENSIONS,
+    ).split(",")
+] + [
```
@RenaudLN Shouldn't we keep these routes private until at least one 'real' route is made public?

RenaudLN commented:
I reverted this PR to just being about refactoring to use the before_request decorator. I will create another PR for the public routes.

alexcjohnson commented:
@RenaudLN this looks great. Can you update the browser-tools orb like in https://github.com/plotly/dash/pull/2603/files? That should get the tests running again.

RenaudLN commented:
Done and tests passed :)

alexcjohnson reviewed (approved):
💃 Thanks very much @RenaudLN - and @fkuschel and @olivier-lacroix thank you for your feedback :)

fkuschel commented:
@RenaudLN , thank you very much for your suggestions!

alexcjohnson commented:
@RenaudLN, @fkuschel, and @olivier-lacroix - my apologies, I had forgotten to create a release for this work! But v2.1.0 is now published.

alexcjohnson commented:
@RenaudLN this PR includes whitelisting, does that mean we can close #142 or is there additional functionality there?

RenaudLN commented:
@alexcjohnson I had split out the whitelisting from this PR as this one was originally about protecting all routes with before_request and whitelisting was starting to muddy up the changes. #142 is still required to add the whitelist functionality.

alexcjohnson commented:
Ah ok, I had forgotten that in the history here. If you want to update that PR I can make another release!

avegancafe commented:
Is there no way anymore to have unprotected domains? This seems like a breaking change
